include/Dialect/Poly/IR/PolyAttributes.td

```diff
@@ -0,0 +1,71 @@
+#ifndef HEIR_INCLUDE_DIALECT_POLY_IR_POLYATTRIBUTES_TD_
+#define HEIR_INCLUDE_DIALECT_POLY_IR_POLYATTRIBUTES_TD_
+
+include "PolyDialect.td"
+
+include "mlir/IR/DialectBase.td"
+include "mlir/IR/AttrTypeBase.td"
+
+class Poly_Attr<string name, string attrMnemonic, list<Trait> traits = []>
+    : AttrDef<Poly_Dialect, name, traits> {
+  let mnemonic = attrMnemonic;
+}
+
+class Poly_Attr_With_Custom_Format<string name, list<Trait> traits = []>
+    : AttrDef<Poly_Dialect, name, traits> {
+  let mnemonic = ?;
+}
+
+def Polynomial_Attr : Poly_Attr<"Polynomial", "polynomial"> {
+  let summary = "An attribute containing a polynomial.";
+  let description = [{
+     #poly = #poly.poly<x**1024 + 1>
+  }];
+
+  let parameters = (ins "Polynomial":$value);
+
+  let builders = [
+    AttrBuilderWithInferredContext<(ins "Polynomial":$value), [{
+      return $_get(value.getContext(), value);
+    }]>
+  ];
+  let extraClassDeclaration = [{
+    using ValueType = Polynomial;
+    Polynomial getPolynomial() const { return getValue(); }
+  }];
+
+  let skipDefaultBuilders = 1;
+  let hasCustomAssemblyFormat = 1;
+}
+
+def Ring_Attr : Poly_Attr<"Ring", "ring"> {
+  let summary = "An attribute specifying a ring.";
+  let description = [{
+    An attribute specifying a polynomial quotient ring with integer
+    coefficients, $\mathbb{Z}/n\mathbb{Z}[x] / (p(x))$.
+
+    `cmod` is the coefficient modulus $n$, and `ideal` is the ring ideal
+    $(p(x))$. Because all ideals in a single-variable polynomial ring are
+    principal, the ideal is defined by a single polynomial.
+
+      #ring = #poly.ring<cmod=1234, ideal=#poly.polynomial<x**1024 + 1>>
+  }];
+
+  let parameters = (ins "APInt": $cmod, "Polynomial":$ideal);
+
+  let builders = [
+    AttrBuilderWithInferredContext<(ins "APInt": $cmod, "Polynomial":$ideal), [{
+      return $_get(ideal.getContext(), cmod, ideal);
+    }]>
+  ];
+  let extraClassDeclaration = [{
```

Comment:
might be nice to have an extra getter for degree of the polynomial? WDYT?

speaking of: because degree is usually determined as `n` maybe the coefficient modulus should be `q` (and we can specificy that it doesn't necessarily require being prime)

Reply:
Do you mean the polynomial's actual degree (which is data dependent; `x^3 + 1` has degree 3 in `Z[x]/(x^8 + 1)`) or just a single-method helper that calls `getPolynomial().getDegree()`? If the latter, I don't see two chained method calls as a bad thing, as it keeps the API surface smaller. Maybe we might want to tighten down the Polynomial API though, and not expose the Monomial list directly....

Reply:
Yeah, here it would be `ideal().getPolynomial().getDegree()` - that's fine! agree let's keep the API and tablegen as minimal as possible.